superset/db_engine_specs/ocient.py

```diff
@@ -215,32 +222,55 @@ class OcientEngineSpec(BaseEngineSpec):
 
     @classmethod
     def get_table_names(
-        cls, database: "Database", inspector: Inspector, schema: Optional[str]
+        cls, database: Database, inspector: Inspector, schema: Optional[str]
     ) -> List[str]:
         return sorted(inspector.get_table_names(schema))
 
     @classmethod
     def fetch_data(cls, cursor, lim=None):
         rows = super(OcientEngineSpec, cls).fetch_data(cursor)
+                
+        if len(rows) > 0 and type(rows[0]).__name__ == rows:
+            # Peek at the schema to determine which column values, if any,
+            # require sanitization.
+            columns_to_sanitize: List[PlacedSanitizeFunc] = _find_columns_to_sanitize(cursor)
 
-        if (len(rows)) == 0:
-            # No rows were produced
-            return rows
+            if columns_to_sanitize:
+                # At least 1 column has to be sanitized.
+                for row in rows:
+                    for info in columns_to_sanitize:
+                        # Modify the element in-place.
+                        v = row[info.column_index]
+                        row[info.column_index] = info.sanitize_func(v)
 
-        if type(rows[0]).__name__ != 'Row':
-            # TODO what else is returned here?
-            return rows
+        # # We are done with this cursor so we can safely remove it from the cache
+        query_ids_to_delete = [q_id for q_id in cursor_cache if cursor_cache[q_id] == cursor]
+        for q_id in query_ids_to_delete:
+            del cursor_cache[q_id]
+        # # We are done with this cursor so we can safely remove it from the cache
+        query_ids_to_delete = [q_id for q_id in cursor_cache if cursor_cache[q_id] == cursor]
+        for q_id in query_ids_to_delete:
+            del cursor_cache[q_id]
+        return rows
 
-        # Peek at the schema to determine which column values, if any,
-        # require sanitization.
-        columns_to_sanitize: List[PlacedSanitizeFunc] = _find_columns_to_sanitize(cursor)
 
-        if columns_to_sanitize:
-            # At least 1 column has to be sanitized.
-            for row in rows:
-                for info in columns_to_sanitize:
-                    # Modify the element in-place.
-                    v = row[info.column_index]
-                    row[info.column_index] = info.sanitize_func(v)
+    @classmethod
+    def get_cancel_query_id(cls, cursor: Any, query: Query) -> Optional[str]:
+        cursor_cache[query.id] = cursor
+        
+        # Return a Non-None value
+        # If None is returned, Superset will not call cancel_query
+        return 'DUMMY_VALUE'
 
-        return rows
+    @classmethod
+    def cancel_query(cls, cursor: Any, query: Query, cancel_query_id: str) -> bool:
+
+        if query.id in cursor_cache:
+            cursor.execute(f'CANCEL {cursor_cache[query.id].query_id}')
+            # Query has been cancelled, so we can safely remove the cursor from the cache
+            del cursor_cache[query.id]
```

Comment:
Also, is this thread safe? Under what context does this stuff get executed? i.e. what happens if two users are canceling queries through the UI?

Reply:
It's per superset instance so yes

Reply:
The spec is reloaded every time the ui is loaded, so each ui instance has their own copy of the spec. It is thread safe

Reply:
I'm confused on how this works at all then... The cache is intended to be stateful across REST calls, no?

Reply:
I.e. "Submit query" and "Cancel query" are two distinct calls to the superset REST API

Reply:
No, the cache is per instance of the webui. You cannot access the queries in progress in other instances of the webui. As such, you will only be able to cancel queries in the existing tab. All tabs will have their own caches

Reply:
I'm still confused, when you say "tab" do you mean a tab in my web browser? 

As in every time I send an HTTP GET request to superset (i.e. port 8088 or w/e) the engine is reloaded into memory. That doesn't sound scalable...

Reply:
Can you point me to the smoking gun for this behavior?

Reply:
(You can sanity check this by creating a nonce
```
nonce = uuid.uuid4()
```
and logging it to the superset console in `fetch_data()` or something, then open two tabs, submit queries can compare the output.

Reply:
I was wrong, there is only one cache. However, the webui only makes the queries from that instance of the UI available to be cancelled. However, the superset handles multiple cancellations gracefully from multiple instances of the webui at once.

Reply:
Well, I think my concern is about how we determine it's handled "gracefully" without knowing the application's threading model. 

What we know:
- `superset.db_engine_specs.ocient..cursor_cache` is not thread safe
- A Superset user or users can issue multiple queries via SQL Lab interface concurrently

What I suspect:
- It seems like an architectural bottleneck to enforce mutual exclusivity amongst these contexts. In other words, it wouldn't be scalable if only 1 user could use the SQL Lab at any given time, right?

How can we guarantee there isn't a race condition on access to `cursor_cache`? And how do we know the thread that calls `fetch_data()` is the same thread that calls `get_cancel_query_id()` _and_ `cancel_query()`?

Reply:
I don't know if multiple users can use it at once. I will need to check